File: backend/app/sim_feed.py
"""Synthetic tick generator for offline development.

Enabled with ``DK_SIMULATE=1``.  It only runs when no SmartAPI session is
active, and the HUD renders a persistent ``SIMULATED FEED`` badge whenever it is
driving the terminal — simulated prints must never be mistaken for live ones.

The model is deliberately simple but structurally faithful: a mean-reverting
random walk on each index, Black-Scholes-ish intrinsic + time value on each
option, and open interest that accumulates around the walls so the COA 2.0
Aegis/Zenith logic has something real to chew on.
"""

from __future__ import annotations

import asyncio
import math
import random
from datetime import datetime

from .config import INDEX_UNIVERSE
from .schemas import Tick
from .scrip_master import ScripMaster
from .ws_manager import TICKS, TickStore

SEED_SPOT = {"NIFTY": 24_500.0, "BANKNIFTY": 52_000.0, "FINNIFTY": 23_400.0}
IV = 0.14


def _premium(spot: float, strike: float, option_type: str, t_years: float) -> float:
    """Cheap approximation: intrinsic + Brownian time value."""
    intrinsic = max(0.0, spot - strike) if option_type == "CE" else max(0.0, strike - spot)
    time_value = spot * IV * math.sqrt(max(t_years, 1e-4)) * 0.4
    moneyness_decay = math.exp(-((spot - strike) ** 2) / (2 * (spot * 0.02) ** 2))
    return round(max(0.05, intrinsic + time_value * moneyness_decay), 2)


class SimulatedFeed:
    """Drives a :class:`TickStore` with plausible index and option prints."""

    def __init__(self, store: TickStore) -> None:
        self.store = store
        self.running = False
        self._task: asyncio.Task | None = None
        self._spot: dict[str, float] = dict(SEED_SPOT)
        self._anchor: dict[str, float] = dict(SEED_SPOT)
        self._oi: dict[str, int] = {}
        self._tick_no = 0

    async def start(self, master: ScripMaster, interval: float = 1.0) -> None:
        await self.stop()
        self.running = True
        self._task = asyncio.create_task(
            self._loop(master, interval), name="sim-feed"
        )

    async def stop(self) -> None:
        self.running = False
        if self._task and not self._task.done():
            self._task.cancel()
            try:
                await self._task
            except (asyncio.CancelledError, Exception):  # noqa: BLE001
                pass
        self._task = None

    async def _loop(self, master: ScripMaster, interval: float) -> None:
        try:
            while self.running:
                self.step(master)
                await asyncio.sleep(interval)
        except asyncio.CancelledError:
            return

    # ------------------------------------------------------------------ #
    def step(self, master: ScripMaster) -> None:
        """Advance the simulation by one frame."""
        self._tick_no += 1
        today = datetime.now().date()

        for underlying, cfg in INDEX_UNIVERSE.items():
            anchor = self._anchor[underlying]
            spot = self._spot[underlying]
            # Ornstein-Uhlenbeck style drift back toward the session anchor.
            spot += (anchor - spot) * 0.02 + random.gauss(0, anchor * 0.0006)
            self._spot[underlying] = spot

            spot_inst = master.spots.get(underlying)
            if spot_inst:
                self.store.apply(
                    Tick(
                        token=spot_inst.token,
                        ltp=round(spot, 2),
                        close=round(anchor, 2),
                        open=round(anchor, 2),
                        high=round(max(spot, anchor), 2),
                        low=round(min(spot, anchor), 2),
                        exchange_ts=int(datetime.now().timestamp() * 1000),
                    )
                )

            expiry = master.nearest_expiry(underlying)
            if not expiry:
                continue
            t_years = max((expiry - today).days, 0) / 365.0

            step = float(cfg["strike_step"])
            for inst in master.contracts(underlying, expiry):
                if abs(inst.strike - spot) > step * 14:
                    continue
                px = _premium(spot, inst.strike, inst.option_type or "CE", t_years)
                px = round(px * random.uniform(0.997, 1.003), 2)

                base = self._oi.get(inst.token)
                if base is None:
                    base = int(abs(random.gauss(0, 1)) * 200_000) + 25_000
                    self._oi[inst.token] = base
                # Writers pile in one step below spot (puts) / above spot (calls).
                pull = math.exp(-abs(inst.strike - spot) / (step * 4))
                growth = int(pull * random.uniform(-400, 1400))
                self._oi[inst.token] = max(1000, self._oi[inst.token] + growth)

                self.store.apply(
                    Tick(
                        token=inst.token,
                        ltp=px,
                        close=round(px * 0.98, 2),
                        volume=int(pull * random.uniform(1e4, 8e4)),
                        oi=self._oi[inst.token],
                        oi_change_pct=round(growth / max(base, 1) * 100, 3),
                        best_bid=round(px * 0.998, 2),
                        best_ask=round(px * 1.002, 2),
                        exchange_ts=int(datetime.now().timestamp() * 1000),
                    )
                )


SIM_FEED = SimulatedFeed(TICKS)

File: web/lib/stream/simFeed.ts
"use client";

import { INDEX_UNIVERSE } from "@/lib/engine/config";
import type { ScripMaster } from "@/lib/engine/scripMaster";
import { type Tick, emptyTick } from "./ticks";

/**
 * Synthetic tick generator — port of `backend/app/sim_feed.py`.
 *
 * Lets the terminal be driven without a SmartAPI session. The HUD renders a
 * persistent SIMULATED FEED badge whenever it is active: synthetic prints must
 * never be mistaken for live ones.
 */

const SEED_SPOT: Record<string, number> = {
  NIFTY: 24_500,
  BANKNIFTY: 52_000,
  FINNIFTY: 23_400,
};
const IV = 0.14;

function gaussian(): number {
  // Box–Muller: Math.random() alone gives a flat distribution, which produces
  // a visibly unnatural tape.
  let u = 0;
  let v = 0;
  while (u === 0) u = Math.random();
  while (v === 0) v = Math.random();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

function premium(spot: number, strike: number, type: string, tYears: number): number {
  const intrinsic = type === "CE" ? Math.max(0, spot - strike) : Math.max(0, strike - spot);
  const timeValue = spot * IV * Math.sqrt(Math.max(tYears, 1e-4)) * 0.4;
  const decay = Math.exp(-((spot - strike) ** 2) / (2 * (spot * 0.02) ** 2));
  return Math.max(0.05, intrinsic + timeValue * decay);
}

export class SimulatedFeed {
  private spot: Record<string, number> = { ...SEED_SPOT };
  private anchor: Record<string, number> = { ...SEED_SPOT };
  private oi = new Map<string, number>();
  private timer: ReturnType<typeof setInterval> | null = null;
  running = false;

  constructor(private onTick: (tick: Tick) => void) {}

  start(master: ScripMaster, intervalMs = 1000): void {
    this.stop();
    this.running = true;
    this.timer = setInterval(() => this.step(master), intervalMs);
    this.step(master);
  }

  stop(): void {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
    this.running = false;
  }

  step(master: ScripMaster): void {
    const today = new Date();

    for (const [underlying, cfg] of Object.entries(INDEX_UNIVERSE)) {
      const anchor = this.anchor[underlying];
      // Ornstein-Uhlenbeck style drift back toward the session anchor.
      let spot = this.spot[underlying];
      spot += (anchor - spot) * 0.02 + gaussian() * anchor * 0.0006;
      this.spot[underlying] = spot;

      const spotToken = master.spotToken(underlying);
      if (spotToken) {
        const t = emptyTick(spotToken);
        t.ltp = Number(spot.toFixed(2));
        t.close = Number(anchor.toFixed(2));
        t.open = t.close;
        t.high = Number(Math.max(spot, anchor).toFixed(2));
        t.low = Number(Math.min(spot, anchor).toFixed(2));
        t.exchangeTs = Date.now();
        this.onTick(t);
      }

      const expiry = master.nearestExpiry(underlying);
      if (!expiry) continue;
      const days = Math.max(
        0,
        (new Date(expiry).getTime() - today.getTime()) / 86_400_000,
      );
      const tYears = days / 365;
      const step = cfg.strikeStep;

      for (const inst of master.contracts(underlying, expiry)) {
        if (Math.abs(inst.strike - spot) > step * 14) continue;

        const px = Number(
          (
            premium(spot, inst.strike, inst.optionType ?? "CE", tYears) *
            (0.997 + Math.random() * 0.006)
          ).toFixed(2),
        );

        let base = this.oi.get(inst.token);
        if (base === undefined) {
          base = Math.floor(Math.abs(gaussian()) * 200_000) + 25_000;
          this.oi.set(inst.token, base);
        }
        // Writers pile in near spot: puts below, calls above.
        const pull = Math.exp(-Math.abs(inst.strike - spot) / (step * 4));
        const growth = Math.floor(pull * (Math.random() * 1800 - 400));
        const next = Math.max(1000, (this.oi.get(inst.token) ?? base) + growth);
        this.oi.set(inst.token, next);

        const t = emptyTick(inst.token);
        t.ltp = px;
        t.close = Number((px * 0.98).toFixed(2));
        t.volume = Math.floor(pull * (1e4 + Math.random() * 7e4));
        t.oi = next;
        t.oiChangePct = Number(((growth / Math.max(base, 1)) * 100).toFixed(3));
        t.bestBid = Number((px * 0.998).toFixed(2));
        t.bestAsk = Number((px * 1.002).toFixed(2));
        t.exchangeTs = Date.now();
        this.onTick(t);
      }
    }
  }
}
